### ISSUE TITLE
User directory does not appear to return users who are not in a public room, regardless of user_directory.search_all_users value

### ISSUE THREAD

**Issue by benparsons** [matrix-org/synapse#11104]
### Description

We added the following settings to homeserver.yaml:

```
user_directory:
  enabled: true
  search_all_users: true
```

Then searched for a user who was not in a shared room or any public room. The targe user was not returned by the search.

### Steps to reproduce

- New instance. Create user_a, user_b.
- Enter the following in homeserver.yaml:

```
user_directory:
  enabled: true
  search_all_users: true
```

- user_a searches for user_b, cannot find them by mxid or display name
- user_b creates a new public room, joins it
- user_a and user_b still do not share a room
- user_a searches for user_b, finds them

### Version information

- **Homeserver**: Synapse 1.27

If not matrix.org: private customer

- **Version**: Synapse 1.27

- **Install method**:

- **Platform**: k8s

**Comment by DMRobertson** [comment]
Up until recently, after setting `search_all_users` on you had to manually rebuild the user directory to include these users. Instructions to do so are in [the second paragraph here](https://matrix-org.github.io/synapse/latest/user_directory.html).

I think any users created after `search_all_users` is set True and Synapse is restarted should show up in the directory, even if they're in zero rooms.

The reproduction steps should not cause the bug as of #10796, which was included in Synapse 1.44.

(Aside: Synapse 1.27 is old and contains known security vulnerabilities.)

**Comment by richvdh** [comment]
assuming this is no longer an issue
